File: tests/src/test/java/client/DummyCommunicationSystem.java
package client;

import com.github.blahblahbloopster.crypto.CommunicationSystem;
import kotlin.Unit;
import kotlin.jvm.functions.Function2;
import org.jetbrains.annotations.NotNull;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/** A dummy {@link CommunicationSystem} for tests. */
public class DummyCommunicationSystem implements CommunicationSystem {
    private final List<Function2<byte[], Integer, Unit>> listeners = new ArrayList<>();
    private final int id = new Random().nextInt();
    private static final ArrayList<DummyCommunicationSystem> systems = new ArrayList<>();

    {
        systems.add(this);
    }

    @NotNull
    @Override
    public List<Function2<byte[], Integer, Unit>> getListeners() {
        return listeners;
    }

    @Override
    public int getId() {
        return id;
    }

    private void received(byte[] bytes, int sender) {
        listeners.forEach(item -> item.invoke(bytes, sender));
    }

    @Override
    public void send(@NotNull byte[] bytes) {
        systems.forEach(item -> {
            if (item != this) {
                item.received(bytes, id);
            }
        });
    }

    @Override
    public void init() {}
}

File: client/src/com/github/blahblahbloopster/crypto/DummyCommunicationSystem.kt
package com.github.blahblahbloopster.crypto

import java.util.*
import java.util.function.*

/** A dummy [CommunicationSystem] for tests.  */
class DummyCommunicationSystem : CommunicationSystem {
    override val listeners: MutableList<(ByteArray, Int) -> Unit> = mutableListOf()
    override val id = Random().nextInt()

    private fun received(bytes: ByteArray, sender: Int) {
        listeners.forEach(Consumer { it.invoke(bytes, sender) })
    }

    override fun send(bytes: ByteArray) {
        systems.forEach(Consumer { item: DummyCommunicationSystem ->
            if (item !== this) {
                item.received(bytes, id)
            }
        })
    }

    override fun init() {}

    companion object {
        private val systems = ArrayList<DummyCommunicationSystem>()
    }

    init {
        systems.add(this)
    }
}